src/colorizer/AnnotationData.ts

```diff
@@ -0,0 +1,222 @@
+  private numLabelsCreated: number;
+  /**
+   * Cached mapping from time to label indices to IDs. Must be invalidated when
+   * labels are removed, or if annotations are applied to or removed from
+   * objects.
+   */
+  private timeToLabelIdMap: Map<number, Record<number, number[]>> | null;
+
+  constructor() {
+    this.labelData = [];
+    this.numLabelsCreated = 0;
+    this.timeToLabelIdMap = null;
+
+    this.getLabelsAppliedToId = this.getLabelsAppliedToId.bind(this);
+    this.getLabeledIds = this.getLabeledIds.bind(this);
+    this.getTimeToLabelIdMap = this.getTimeToLabelIdMap.bind(this);
+    this.getLabels = this.getLabels.bind(this);
+    this.createNewLabel = this.createNewLabel.bind(this);
+    this.setLabelName = this.setLabelName.bind(this);
+    this.setLabelColor = this.setLabelColor.bind(this);
+    this.deleteLabel = this.deleteLabel.bind(this);
+    this.toggleLabelOnId = this.toggleLabelOnId.bind(this);
+    this.applyLabelToId = this.applyLabelToId.bind(this);
+    this.removeLabelFromId = this.removeLabelFromId.bind(this);
+  }
+
+  // Getters
+
+  getLabels(): LabelData[] {
+    return [...this.labelData];
+  }
+
+  getLabelsAppliedToId(id: number): number[] {
+    const labelIdxs: number[] = [];
+    for (let i = 0; i < this.labelData.length; i++) {
+      if (this.labelData[i].ids.has(id)) {
+        labelIdxs.push(i);
+      }
+    }
+    return labelIdxs;
+  }
+
+  getLabeledIds(labelIdx: number): number[] {
+    this.validateIndex(labelIdx);
+    return Array.from(this.labelData[labelIdx].ids);
+  }
+
+  getTimeToLabelIdMap(dataset: Dataset): Map<number, Record<number, number[]>> {
+    if (this.timeToLabelIdMap !== null) {
+      return this.timeToLabelIdMap;
+    }
+
+    const timeToLabelIdMap = new Map<number, Record<number, number[]>>();
+
+    for (let labelIdx = 0; labelIdx < this.labelData.length; labelIdx++) {
+      const ids = this.labelData[labelIdx].ids;
+      for (const id of ids) {
+        const time = dataset.times?.[id];
+        if (time === undefined) {
+          continue;
+        }
+        if (!timeToLabelIdMap.has(time)) {
+          timeToLabelIdMap.set(time, {});
+        }
+        if (!timeToLabelIdMap.get(time)![labelIdx]) {
+          timeToLabelIdMap.get(time)![labelIdx] = [];
+        }
+        timeToLabelIdMap.get(time)![labelIdx].push(id);
+      }
+    }
+    this.timeToLabelIdMap = timeToLabelIdMap;
+    return timeToLabelIdMap;
+  }
+
+  // Setters
+
+  /** Creates a new label and returns its index. */
+  createNewLabel(name?: string, color?: Color): number {
+    if (!color) {
+      const palette = KNOWN_CATEGORICAL_PALETTES.get(DEFAULT_CATEGORICAL_PALETTE_KEY)!;
+      color = new Color(palette.colorStops[this.numLabelsCreated % palette.colorStops.length]);
+    }
+    if (!name) {
+      name = `Label ${this.numLabelsCreated + 1}`;
+    }
+
+    this.labelData.push({
+      name: name,
+      color: color,
+      ids: new Set(),
+    });
```

Comment:
Optional: can use implicit names

```suggestion
    this.labelData.push({
      name,
      color,
      ids: new Set(),
    });
```